iothub/device/src/Pipeline/ConnectionStateDelegatingHandler.cs

```diff
@@ -0,0 +1,625 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    internal sealed class ConnectionStateDelegatingHandler : DefaultDelegatingHandler
+    {
+        private readonly SemaphoreSlim _clientOpenSemaphore = new(1, 1);
+
+        private readonly Action<ConnectionStatusInfo> _onConnectionStatusChanged;
+        private readonly IIotHubClientRetryPolicy _retryPolicy;
+
+        private readonly ClientTransportStateMachine _clientTransportStateMachine = new();
+        private Task _transportClosedTask;
+        private CancellationTokenSource _handleDisconnectCts;
+        private CancellationTokenSource _cancelPendingOperationsCts;
+
+        internal ConnectionStateDelegatingHandler(PipelineContext context, IDelegatingHandler innerHandler)
+            : base(context, innerHandler)
+        {
+            _onConnectionStatusChanged = context.ConnectionStatusChangeHandler;
+
+            // This retry policy is saved only to handle the client state when the retry policy informs that no more retry attempts are to be made
+            _retryPolicy = context.RetryPolicy;
```

Comment:
Todo - check exp backoff retry policy. Possible that we increase the wait time way quicker than "required"